cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -0,0 +1,455 @@
+                  page_offsets.emplace_back(page_offsets.back() + page_counts.back());
+                }
+              });
+            });
+          return std::pair{page_counts, page_offsets};
+        }();
+
+        auto const num_pages = page_offsets.back();
+        CUDF_EXPECTS(num_pages > 0, "No pages with PageIndex found for the column");
+
+        host_column<T> min(num_pages, stream);
+        host_column<T> max(num_pages, stream);
+
+        // For each source
+        std::for_each(
+          thrust::counting_iterator<size_t>(0),
+          thrust::counting_iterator(row_group_indices.size()),
+          [&, page_offsets = page_offsets](auto src_idx) {
+            // For all row groups in this source
+            auto const& rg_indices = row_group_indices[src_idx];
+            std::for_each(rg_indices.cbegin(), rg_indices.cend(), [&](auto rg_idx) {
+              auto const& row_group = per_file_metadata[src_idx].row_groups[rg_idx];
+              auto col =
+                std::find_if(row_group.columns.begin(),
+                             row_group.columns.end(),
+                             [schema_idx](cudf::io::parquet::detail::ColumnChunk const& col) {
+                               return col.schema_idx == schema_idx;
+                             });
+              if (col != std::end(row_group.columns) and col->column_index.has_value()) {
+                auto const& colchunk             = *col;
+                auto const& column_index         = colchunk.column_index.value();
+                auto const& offset_index         = colchunk.offset_index.value();
+                auto const num_pages_in_colchunk = column_index.min_values.size();
+                auto const page_offset_in_colchunk =
+                  page_offsets[src_idx * row_group_indices.size() + rg_idx];
+
+                // For all pages in this chunk
+                std::for_each(
+                  thrust::counting_iterator<size_t>(0),
+                  thrust::counting_iterator(num_pages_in_colchunk),
+                  [&](auto page_idx) {
+                    // To support deprecated min, max fields.
+                    auto const& min_value = column_index.min_values[page_idx];
+                    auto const& max_value = column_index.min_values[page_idx];
+                    // translate binary data to Type then to <T>
+                    min.set_index(
+                      page_offset_in_colchunk + page_idx, min_value, colchunk.meta_data.type);
+                    max.set_index(
+                      page_offset_in_colchunk + page_idx, max_value, colchunk.meta_data.type);
+                  });
+              }
+            });
+          });
+
+        auto mincol = min.to_device(dtype, stream, mr);
+        auto maxcol = max.to_device(dtype, stream, mr);
+
+        // auto null_mask = std::move(contents.null_mask);
+        auto counts  = cudf::detail::make_device_uvector_async(page_counts, stream, mr);
```

Comment:
Should be the count of rows in each page. Likewise, offsets should be the starting row indices of each page.